Allow smbd to load libcephfs proxy conditionally

anoopcs9 opened the pull request:
The design for [libcephfs proxy](https://docs.ceph.com/en/latest/dev/libcephfs_proxy/) was accepted sometime back with the corresponding implementation subsequently [merged](https://github.com/ceph/ceph/pull/58376) into upstream Ceph. This additional way to integrate CephFS with Samba is comprised of two components. The daemon i.e, libcephfsd is responsible for creating the unix domain socket as a starting point of communication. On the other side proxy library, which is [designed](https://gitlab.com/samba-team/samba/-/merge_requests/3792) to be loaded within _smbd_, connects to the already created socket and thereby communicates with the daemon.

During the development and testing stages we happened to notice some AVC denial messages in the audit logs:

> type=AVC msg=audit(1742367267.967:6919): avc:  denied  { write } for  pid=62992
  comm="smbd[192.168.12" name="libcephfsd.sock" dev="tmpfs" ino=6049
  scontext=system_u:system_r:smbd_t:s0 tcontext=unconfined_u:object_r:var_run_t:s0
  tclass=sock_file permissive=1

> type=AVC msg=audit(1742367267.967:6919): avc:  denied  { connectto } for  pid=62992
  comm="smbd[192.168.12" path="/run/libcephfsd.sock" scontext=system_u:system_r:smbd_t:s0
  tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
  tclass=unix_stream_socket permissive=1

> type=AVC msg=audit(1742367267.970:6920): avc:  denied  { read write } for  pid=62992
  comm="smbd[192.168.12" path="socket:[189595]" dev="sockfs" ino=189595 scontext=system_u:system_r:smbd_t:s0
  tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
  tclass=unix_stream_socket permissive=1

Therefore we propose a new boolean - **_samba_load_libcephfs_proxy_**, off by default, to contain the necessary allow rules for _smbd_t_ to operate in a proxy mode using libcephfs.

anoopcs9 committed:
Allow smbd to load libcephfs proxy conditionally

The design for libcephfs proxy[1] was accepted sometime back with the
corresponding implementation subsequently merged[2] into upstream Ceph.
This additional way to integrate CephFS with Samba is comprised of two
components. The daemon i.e, libcephfsd is responsible for creating the
unix domain socket as a starting point of communication. On the other
side proxy library, which is designed[3] to be loaded within smbd,
connects to the already created socket and thereby communicates with
the daemon.

During the development and testing stages we happened to notice some AVC
denial messages in the audit logs:

type=AVC msg=audit(1742367267.967:6919): avc:  denied  { write } for  pid=62992
  comm="smbd[192.168.12" name="libcephfsd.sock" dev="tmpfs" ino=6049
  scontext=system_u:system_r:smbd_t:s0 tcontext=unconfined_u:object_r:var_run_t:s0
  tclass=sock_file permissive=1

type=AVC msg=audit(1742367267.967:6919): avc:  denied  { connectto } for  pid=62992
  comm="smbd[192.168.12" path="/run/libcephfsd.sock" scontext=system_u:system_r:smbd_t:s0
  tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
  tclass=unix_stream_socket permissive=1

type=AVC msg=audit(1742367267.970:6920): avc:  denied  { read write } for  pid=62992
  comm="smbd[192.168.12" path="socket:[189595]" dev="sockfs" ino=189595 scontext=system_u:system_r:smbd_t:s0
  tcontext=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
  tclass=unix_stream_socket permissive=1

Therefore we propose a new boolean - samba_load_libcephfs_proxy, off by
default, to contain the necessary allow rules for smbd_t to operate in a
proxy mode using libcephfs.

[1] https://docs.ceph.com/en/latest/dev/libcephfs_proxy/
[2] https://github.com/ceph/ceph/pull/58376
[3] https://gitlab.com/samba-team/samba/-/merge_requests/3792

Signed-off-by: Anoop C S <anoopcs@cryptolab.net>

anoopcs9 commented:
@zpytela I hope **rpm-build:fedora-41-x86_64** job failure is expected.

anoopcs9 commented:
@zpytela Can you please take a look at this change?

zpytela commented on policy/modules/contrib/samba.te:
```diff
@@ -584,6 +591,13 @@ tunable_policy(`samba_load_libgfapi',`
+optional_policy(`
+    tunable_policy(`samba_load_libcephfs_proxy',`
+        unconfined_stream_connect(smbd_t)
+        rw_sock_files_pattern(smbd_t, var_run_t, var_run_t)
```
Which service creates and/or works with files with the `var_run_t` type? A private type should be used in cases of allowing write access.

anoopcs9 replied:
Proxy daemon, _libcephfsd_, creates the socket file for read/write under _/run_. It is a standalone daemon outside _smbd_. The proxy client library loaded by _smbd_ connects to the socket post the creation by _libcephfsd_.

zpytela replied:
There is no service in libcephsd, which service starts/uses it?
Which filename is used?

Edit: If it was this one, the label should be smbd_var_run_t
command: systemctl set-environment LIBCEPHFSD_SOCKET=/run/samba/libcephfsd.sock

anoopcs9 replied:
The catch here is that the socket file created by _/usr/sbin/libcephfsd_ can be used by any application designed to work with the proxy client library. IOW, Samba may not be the only consumer of the socket file created by _/usr/sbin/libcephfsd_ and thus it creates under _/run_.

Currently there is no service responsible for launching the daemon(_libcephfsd_). It is supposed to be running before any consumer attempts to communicate via the socket file.

anoopcs9 committed:
Add unconfined_stream_connect_rw interface

When ancillary data is used for passing file descriptors via unix domain
sockets, SELinux hook for file_receive inside kernel performs additional
permission checks on each of those file descriptors received and their
respective underlying inode. For socket file descriptors(AF_UNIX) passed
around this would mean the source comes with read+write permissions on
unix_stream_socket which otherwise result in injecting MSG_CTRUNC flag
while returning recvmsg() without failing the system call itself. In the
end the ancillary data is lost defeating the whole purpose. Thus we
introduce a new interface to tackle the additional read+write premission
in general as an extension to the existing 'unconfined_stream_connect'
interface.

Please note that the denial itself is not logged for systemprocess and
daemon with the help of 'unconfined_dontaudit_rw_stream' interface. They
are visible once whole policy is rebuilt after removing dontaudit rules
using `semodule -DB`.

Signed-off-by: Anoop C S <anoopcs@cryptolab.net>

anoopcs9 commented:
I've updated the PR with an additional commit(explaining the situation) as we discovered a hidden denial breaking the communication between proxy client and daemon since https://github.com/ceph/ceph/pull/61877.

anoopcs9 commented:
@zpytela How do we plan to take this forward?

zpytela commented:
Services should be allowed write access to their private types, not to generic ones. If a socket file is created by some service, the service needs to be confined, too.

anoopcs9 commented:
> Services should be allowed write access to their private types, not to generic ones. If a socket file is created by some service, the service needs to be confined, too.

Ok, this would mean we create the required types for ceph related services which is more involved (although I see the necessary [content](https://github.com/ceph/ceph/tree/main/selinux) in upstream). I'll come back once ceph contexts and types are in place.
